app/src/main/java/com/example/habitsmasher/UserRegistrationFragment.java

```diff
@@ -0,0 +1,213 @@
+        _auth.createUserWithEmailAndPassword(email, password)
+             .addOnCompleteListener(new OnCompleteListener<AuthResult>() {
+                 @Override
+                 public void onComplete(@NonNull Task<AuthResult> task) {
+                     if (task.isSuccessful()) {
+                         User user = new User(_auth.getUid(),
+                                              username,
+                                              email,
+                                              password);
+
+                         addNewUserToDatabase(user);
+
+                     } else {
+                         showMessage(FAILED_TO_REGISTER_MESSAGE);
+                     }
+                     _progressBar.setVisibility(View.GONE);
+                 }
+             });
+    }
+
+    /**
+     * This method adds a new user to the Firestore database
+     * @param user the new user to add
+     */
+    private void addNewUserToDatabase(User user) {
+        FirebaseFirestore.getInstance()
+                         .collection("Users")
+                         .document(user.getId())
+                         .set(buildUserDataMap(user))
+                         .addOnCompleteListener(new OnCompleteListener<Void>() {
+                             @Override
+                             public void onComplete(@NonNull Task<Void> task) {
+                                 if (task.isSuccessful()) {
+                                     showMessage(USER_REGISTERED_MESSAGE);
+                                     goBackToLoginScreen();
+                                   } else {
+                                     showMessage(FAILED_TO_ADD_USER_MESSAGE);
+                                   }
+                             }
+                         });
+    }
+
+    /**
+     * This helper method shows a toast message to the screen
+     * @param message message to display
+     */
+    private void showMessage(String message) {
+        Toast.makeText(getContext(),
+                       message, Toast.LENGTH_LONG).show();
+    }
+
+    /**
+     * This helper method builds up the user data to insert into the database
+     * @param user the user data to build up
+     * @return a hashmap of key/value pairs of the user data
+     */
+    @NonNull
+    private HashMap<String, Object> buildUserDataMap(User user) {
+        HashMap<String, Object> userData = new HashMap<>();
+
+        userData.put("username", user.getUsername());
+        userData.put("email", user.getEmail());
+        userData.put("password", user.getPassword());
+        userData.put("id", user.getId());
+
+        return userData;
+    }
+
+    /**
+     * This method sets the click listener for the "go back" button and routes to the new fragment
+     * @param backToLoginButton
+     */
+    private void setClickListenerForBackToLoginButton(Button backToLoginButton) {
```

Comment:
Same thing as other comments, Button in parameters

Reply:
I think we can just pick **one** way and stick with it. Right now, I'm just choosing to **initialize** beforehand and pass the reference to the button to set the listener.

This should be fine.

The other way you're suggesting is equally correct, but it's not the biggest deal, to be honest, and I don't want to lose the current approvals.